lib/Core/Memory.cpp

```diff
@@ -624,44 +628,92 @@ void ObjectState::print() const {
 
 /***/
 
-PersistentState::PersistentState(const ObjectState *os, const Array *cacheLines)
+PersistentState::PersistentState(const ObjectState *os, 
+                                 const Array *cacheLines,
+                                 const Array *rootCauses,
+                                 const Array *idxArray)
   : ObjectState(*os),
-    cacheLineUpdates(0, 0),
-    pendingCacheLineUpdates(0, 0) {
-  cacheLineUpdates = UpdateList(cacheLines, 0);
-  pendingCacheLineUpdates = UpdateList(cacheLineUpdates);
+    cacheLineUpdates(cacheLines, nullptr),
+    pendingCacheLineUpdates(cacheLineUpdates),
+    idxEmptyUpdates(idxArray, nullptr),
+    rootCauseLocations(rootCauses, nullptr) {
+  // -- For cacheline stuff
+  // 0 <= X <= 2^32
+  idxUnbounded = ReadExpr::create(idxEmptyUpdates,
+                                  ConstantExpr::create(0, idxArray->range));
+  // idxUnbounded = ReadExpr::createTempRead(idxArray, idxArray->range);
+  // idxUnbounded->dump();
+
+  assert(!(numCacheLines() % nrLinesPerSlice) && "not divisible!");
+
+  #if 0
+  for (unsigned start = 0; start < numCacheLines(); start += nrLinesPerSlice) {
+    ref<Expr> lowerBound = UleExpr::create(ConstantExpr::create(start, idxArray->range), idxUnbounded);
+    unsigned end = start + nrLinesPerSlice;
+    ref<Expr> upperBound = UltExpr::create(idxUnbounded, ConstantExpr::create(end, idxArray->range));
+    ref<Expr> bounds = AndExpr::create(lowerBound, upperBound);
+
+    idxConstraints.push_back(bounds);
+  }
+  #else 
+  ref<Expr> lowerBound = UleExpr::create(ConstantExpr::create(0, idxArray->range), idxUnbounded);
+  ref<Expr> upperBound = UltExpr::create(idxUnbounded, ConstantExpr::create(numCacheLines(), idxArray->range));
+  idxConstraints.push_back(AndExpr::create(lowerBound, upperBound));
+  #endif
```

Comment:
How were you able to create one constraint here rather than at the cacheline granularity?

Reply:
Reading from a symbolic array that has no writes to it (`idxUnbounded`) produces an expression that can equal any value in range of its size, (i.e., if I read 4 bytes from an empty array, it yields an expression that represents any value from [0, 2^32)). By adding the explicit upper and lower bound, this creates an expression that represents any cacheline.